src/transformers/generation_utils.py

```diff
@@ -491,11 +491,21 @@ def generate(
                 raise ValueError("Make sure that `model_kwargs` include `encoder_outputs` of type `ModelOutput`.")
 
         # determine generation mode
+        if num_beams < 1:
+            raise ValueError("num_beams must be >= 1")
         is_greedy_gen_mode = (num_beams == 1) and do_sample is False
         is_sample_gen_mode = (num_beams == 1) and do_sample is True
         is_beam_gen_mode = (num_beams > 1) and do_sample is False
         is_beam_sample_gen_mode = (num_beams > 1) and do_sample is True
 
+        # warn if certain params will be unused
+        if not do_sample:
```

Comment:
I'm happy however to add a statement for each input param to the doc. E.g. under top_p. Note that top_p is only effective if `do_sample=True`

Reply:
Ideas:
- What if we also do it for the other params? In that case what would those other params be? I'm happy to add checks for the other params in this function (though that could mean that this function's behavior will differ from other behaviors in the library since warnings are not generally issued for unused parameters. I do think that's one of the weaknesses of a library this general. It's sometimes hard to find where something isn't behaving as expected because an arg gets dropped or is set unexpectedly in another location.)

- A more involved idea would be to group the params as dicts for each type of generation.

But more generally I agree with you that the user should know that if he plays with top_p, top_k it's only valid if do_sample is true. My problem was that in my wandb sweep code, I had forgotten to set do_sample to True, so I was iterating through a ton of topp, topk settings but seeing no changes in output. At first I thought beam_sample was broken. It took me a while to figure out that my do_sample setting to false was nullifying all the other params I was passing in. A warning would have been nice...

Reply:
In my opinion, it's not worth it to clutter the code for such warnings. Another problem is that this warning will not catch if people change `top_k` and `top_p` with their config and this is also hard to check since `config.top_k` for example is set to 50 already. I'd really prefer to just add those statements to the doc.
If we go down the road of adding warnings for every single parameter, it won't become maintainable, *e.g.* we might merge this PR soon: https://github.com/huggingface/transformers/pull/8627

It's super useful to add these statements to the doc, but not worth the if-else cascade for the warnings IMO. There is already a lot of stuff happening under-the-hood in `generate()` and if one wants to change the default generate configs of the model he/she should be aware of what is going on exactly.